lib/elixir/src/elixir_errors.erl

```diff
@@ -14,89 +16,91 @@
+  print_warning_no_diagnostic(PrintMessage).
 
--spec form_error(list(), binary() | #{file := binary(), _ => _}, module(), any()) -> no_return().
-form_error(Meta, #{file := File}, Module, Desc) ->
-  compile_error(Meta, File, Module:format_error(Desc));
-form_error(Meta, File, Module, Desc) ->
-  compile_error(Meta, File, Module:format_error(Desc)).
+%% Compilation error/warn handling.
 
--spec form_warn(list(), binary() | #{file := binary(), _ => _}, module(), any()) -> ok.
-form_warn(Meta, File, Module, Desc) when is_list(Meta), is_binary(File) ->
-  form_warn(Meta, #{file => File}, Module, Desc);
-form_warn(Meta, #{file := File} = E, Module, Desc) when is_list(Meta) ->
+-spec file_warn(list(), binary() | #{file := binary(), _ => _}, module(), any()) -> ok.
+file_warn(Meta, File, Module, Desc) when is_list(Meta), is_binary(File) ->
+  file_warn(Meta, #{file => File}, Module, Desc);
+file_warn(Meta, #{file := File} = E, Module, Desc) when is_list(Meta) ->
   % Skip warnings during bootstrap, they will be reported during recompilation
   case elixir_config:is_bootstrap() of
     true -> ok;
-    false -> do_form_warn(Meta, File, E, Module:format_error(Desc))
+    false ->
+      {Line, File, Location} = env_format(Meta, E),
+      Warning = Module:format_error(Desc),
+      print_warning(Line, File, Warning, [Warning, "\n  ", Location, $\n])
   end.
 
-do_form_warn(Meta, GivenFile, E, Warning) ->
-  [{file, File}, {line, Line}] = meta_location(Meta, GivenFile),
-
-  Location =
-    case E of
-      #{function := {Name, Arity}, module := Module} ->
-        [file_format(Line, File), ": ", 'Elixir.Exception':format_mfa(Module, Name, Arity)];
-      #{module := Module} when Module /= nil ->
-        [file_format(Line, File), ": ", elixir_aliases:inspect(Module)];
-      #{} ->
-        file_format(Line, File)
-    end,
-
-  log_and_print_warning(Line, File, Warning, [Warning, "\n  ", Location, $\n]).
+-spec file_error(list(), binary() | #{file := binary(), _ => _}, module(), any()) -> no_return().
+file_error(Meta, File, Module, Desc) when is_list(Meta), is_binary(File) ->
+  file_error(Meta, #{file => File}, Module, Desc);
+file_error(Meta, Env, Module, Desc) when is_list(Meta) ->
+  print_error(Meta, Env, Module, Desc),
+  compile_error(Env).
+
+%% A module error is one where it can continue if there is a module
+%% being compiled. If there is no module, it is a regular file_error.
+-spec module_error(list(), #{file := binary(), module => module() | nil, _ => _}, module(), any()) -> ok.
+module_error(Meta, #{module := EnvModule} = Env, Module, Desc) when EnvModule /= nil ->
+  print_error(Meta, Env, Module, Desc),
+  case elixir_module:taint(EnvModule) of
+    true -> ok;
+    false -> compile_error(Env)
+  end;
+module_error(Meta, Env, Module, Desc) ->
+  file_error(Meta, Env, Module, Desc).
+
+%% A function error is one where it can continue if there is a function
+%% being compiled. If there is no function, it is falls back to module_error.
```

Comment:
Maybe I misunderstand, but based on the current implementation it seems to be
```diff
-%% being compiled. If there is no function, it is falls back to module_error.
+%% being compiled. If there is no function, it is falls back to file_error.
```

Should the fallback be to `module_error`?

Reply:
Yes, good catch, I will fix.